src/tools/rbd_mirror/NamespaceReplayer.cc

```diff
@@ -0,0 +1,792 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#include "NamespaceReplayer.h"
+#include <boost/bind.hpp>
+#include "common/Formatter.h"
+#include "common/debug.h"
+#include "common/errno.h"
+#include "cls/rbd/cls_rbd_client.h"
+#include "librbd/Utils.h"
+#include "librbd/api/Config.h"
+#include "librbd/api/Mirror.h"
+#include "ServiceDaemon.h"
+#include "Threads.h"
+
+#define dout_context g_ceph_context
+#define dout_subsys ceph_subsys_rbd_mirror
+#undef dout_prefix
+#define dout_prefix *_dout << "rbd::mirror::NamespaceReplayer: " \
+                           << this << " " << __func__ << ": "
+
+using librbd::util::create_async_context_callback;
+using librbd::util::create_context_callback;
+
+namespace rbd {
+namespace mirror {
+
+using ::operator<<;
+
+namespace {
+
+const std::string SERVICE_DAEMON_INSTANCE_ID_KEY("instance_id");
+const std::string SERVICE_DAEMON_LOCAL_COUNT_KEY("image_local_count");
+const std::string SERVICE_DAEMON_REMOTE_COUNT_KEY("image_remote_count");
+
+} // anonymous namespace
+
+template <typename I>
+NamespaceReplayer<I>::NamespaceReplayer(
+    const std::string &name,
+    librados::IoCtx &local_io_ctx, librados::IoCtx &remote_io_ctx,
+    const std::string &local_mirror_uuid, const std::string &remote_mirror_uuid,
+    Threads<I> *threads, ImageSyncThrottler<I> *image_sync_throttler,
+    ServiceDaemon<I> *service_daemon,
+    journal::CacheManagerHandler *cache_manager_handler) :
+  m_local_mirror_uuid(local_mirror_uuid),
+  m_remote_mirror_uuid(remote_mirror_uuid),
+  m_threads(threads), m_image_sync_throttler(image_sync_throttler),
+  m_service_daemon(service_daemon),
+  m_cache_manager_handler(cache_manager_handler),
+  m_lock(ceph::make_mutex(librbd::util::unique_lock_name(
+      "rbd::mirror::NamespaceReplayer " + name, this))),
+  m_local_pool_watcher_listener(this, true),
+  m_remote_pool_watcher_listener(this, false),
+  m_image_map_listener(this) {
+  dout(10) << name << dendl;
+
+  m_local_io_ctx.dup(local_io_ctx);
+  m_local_io_ctx.set_namespace(name);
+  m_remote_io_ctx.dup(remote_io_ctx);
+  m_remote_io_ctx.set_namespace(name);
+}
+
+template <typename I>
+bool NamespaceReplayer<I>::is_blacklisted() const {
+  std::lock_guard locker{m_lock};
+  return (m_local_pool_watcher &&
+          m_local_pool_watcher->is_blacklisted()) ||
+         (m_remote_pool_watcher &&
+          m_remote_pool_watcher->is_blacklisted());
+}
+
+template <typename I>
+std::string NamespaceReplayer<I>::get_instance_id() const {
```

Comment:
Since the instance id will be the same across all namespaces (since they share the same librados connection), perhaps we can move this logic up to `PoolReplayer` after it connects to the local cluster and then it passes the instance id where it needs to go.

Reply:
@dillaman Fixed this partially: I removed `NamespaceReplayer<I>::get_instance_id` as it was used only in `PoolReplayer::print_status`, where I can temporary just use `m_local_io_ctx.get_instance_id()`.

I agree it is a good idea to have only one authority place for `instance_id` (in `PoolReplayer`) and pass it where it is needed, but this affects many modules not related to this PR, so I think it would be better to do this cleanup separately, in order no to make the review harder.